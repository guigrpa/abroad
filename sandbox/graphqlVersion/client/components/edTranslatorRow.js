// @flow

/* eslint-disable react/no-multi-comp */

import timm from 'timm';
import React from 'react';
import Relay, { graphql } from 'react-relay';
import { cancelEvent, flexContainer, flexItem, Icon, hoverable } from 'giu';
import type { KeyT, HoverableProps } from '../../common/types';
import _t from '../../translate';
import updateKey from '../mutations/updateKey';
import { COLORS } from '../gral/constants';
import { mutate } from './helpers';
import Translation from './eeTranslation';

// ==========================================
// Component declarations
// ==========================================
type PublicProps = {|
  langs: Array<string>,
  fSelected: boolean,
  fPendingSeq: boolean,
  changeSelectedKey: (keyId: ?string) => void,
  styleKeyCol: Object,
  styleLangCol: Object,
  // Relay
  relay: Object,
  theKey: KeyT,
|};

type Props = {
  ...PublicProps,
  ...$Exact<HoverableProps>,
};

const gqlFragments = graphql`
  fragment edTranslatorRow_theKey on Key {
    id
    context
    text
    unusedSince
    ...eeTranslation_theKey
    translations(first: 100000)
      @connection(key: "TranslatorRow_theKey_translations") {
      edges {
        node {
          id
          isDeleted
          lang
          ...eeTranslation_translation
        }
      }
    }
  }
`;

// ==========================================
// Component
// ==========================================
class TranslatorRow extends React.PureComponent {
  props: Props;

  // ------------------------------------------
  // Render
  // ------------------------------------------
  render() {
    const { theKey: key } = this.props;
    const fUnused = !!key.unusedSince;
    const elContext = key.context ? (
      <span style={style.context}>{key.context}</span>
    ) : (
      undefined
    );
    const elText = <span style={style.text}>{key.text}</span>;
    const elDeleteKey = this.props.hovering ? (
      <Icon
        icon="remove"
        title={_t('tooltip_Delete message (does NOT delete any translations)')}
        onClick={this.onClickDeleteKey}
        style={style.removeIcon}
      />
    ) : (
      undefined
    );
    let cellStyle = timm.merge(
      style.bodyCell(this.props.fPendingSeq),
      this.props.styleKeyCol,
      style.keyCell
    );
    const { fSelected } = this.props;
    if (fSelected) cellStyle = style.selected(cellStyle);
    if (fUnused) cellStyle = style.unused(cellStyle);
    return (
      <div
        className="tableBodyRow"
        id={key.id}
        onClick={this.onClickKeyRow}
        style={style.row(this.props)}
      >
        <div
          onMouseEnter={this.props.onHoverStart}
          onMouseMove={this.props.onHoverStart}
          onMouseLeave={this.props.onHoverStop}
          style={cellStyle}
        >
          {elContext}
          {elText}
          {elDeleteKey}
        </div>
        {this.props.langs.map(this.renderTranslation)}
      </div>
    );
  }

  renderTranslation = (lang: string) => {
    const { theKey: key, fPendingSeq } = this.props;
    const edge = key.translations.edges.find(
      ({ node }) => node.lang === lang && !node.isDeleted
    );
    const translation = edge ? edge.node : null;
    const fUnused = !!key.unusedSince;
    let cellStyle = timm.merge(
      style.bodyCell(fPendingSeq),
      this.props.styleLangCol
    );
    if (!edge && !fUnused) cellStyle = style.untranslated(cellStyle);
    if (this.props.fSelected) cellStyle = style.selected(cellStyle);
    return (
      <div key={lang} style={cellStyle}>
        <Translation
          theKey={key}
          lang={lang}
          translation={translation}
          changeSelectedKey={this.props.changeSelectedKey}
        />
      </div>
    );
  };

  // ------------------------------------------
  // Handlers
  // ------------------------------------------
  onClickKeyRow = () => {
    this.props.changeSelectedKey(this.props.theKey.id);
  };

  onClickDeleteKey = (ev: SyntheticKeyboardEvent) => {
    cancelEvent(ev);
    if (this.props.fSelected) {
      this.props.changeSelectedKey(null);
    }
    mutate({
      description: 'Click on Delete key',
      environment: this.props.relay.environment,
      mutationOptions: updateKey({
        theKey: this.props.theKey,
        attrs: { isDeleted: true },
      }),
    });
  };
}

// ------------------------------------------
// Styles
// ------------------------------------------
const style = {
  row: fSelected =>
    flexItem(
      'none',
      flexContainer('row', {
        minHeight: 21,
        maxHeight: fSelected ? undefined : 100,
        overflowY: fSelected ? undefined : 'hidden',
      })
    ),
  bodyCell: fPendingSeq => ({
    position: 'relative',
    paddingTop: 1,
    paddingBottom: 1,
    borderBottom: fPendingSeq ? `1px dashed white` : `1px solid ${COLORS.dark}`,
  }),
  keyCell: {
    paddingRight: 17,
  },
  selected: base =>
    timm.merge(base, {
      backgroundColor: COLORS.medium,
    }),
  unused: base =>
    timm.merge(base, {
      color: COLORS.dim,
    }),
  untranslated: base =>
    timm.merge(base, {
      backgroundColor: COLORS.mediumAlt,
    }),
  context: {
    fontWeight: 900,
    marginRight: 10,
  },
  text: {
    whiteSpace: 'pre-wrap',
  },
  removeIcon: {
    position: 'absolute',
    top: 3,
    right: 5,
    color: 'black',
  },
};

// ==========================================
// Public API
// ==========================================
const HoverableTranslatorRow = hoverable(TranslatorRow);
const Container = Relay.createFragmentContainer(
  HoverableTranslatorRow,
  gqlFragments
);

// Wrap the Relay container in a PureComponent,
// so it only gets rendered when necessary (e.g. not when
// changing the selected row)
class SuperContainer extends React.PureComponent {
  render() {
    return <Container {...this.props} />;
  }
}
export default SuperContainer;
export { HoverableTranslatorRow as _HoverableTranslatorRow };
