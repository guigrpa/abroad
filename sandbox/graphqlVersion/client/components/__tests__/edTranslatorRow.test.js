/* eslint-env jest */
/* eslint-disable global-require */
import React from 'react';
import renderer from 'react-test-renderer';
import { flexItem } from 'giu';
import { _HoverableTranslatorRow as HoverableTranslatorRow } from '../edTranslatorRow';
import {
  VIEWER,
  KEY_WITHOUT_TRANSLATIONS,
  KEY_WITH_TRANSLATIONS,
} from './fixtures';

// https://github.com/facebook/react/issues/7386#issuecomment-238091398
jest.mock('react-dom');
jest.mock('../eeTranslation', () => require('./mockComponent')('Translation'));

// ======================================================
// Fixtures
// ======================================================
const STYLE_KEY_COL = flexItem('1 1 0px', {
  backgroundColor: 'white',
  marginRight: 5,
  paddingLeft: 5,
  paddingRight: 17,
});

const STYLE_LANG_COL = flexItem('1 1 0px', {
  backgroundColor: 'white',
  marginRight: 5,
  paddingLeft: 5,
  paddingRight: 5,
});

// ======================================================
// Tests
// ======================================================
describe('HoverableTranslatorRow', () => {
  it('01 key without translations', () => {
    const tree = renderer
      .create(
        <HoverableTranslatorRow
          theKey={KEY_WITHOUT_TRANSLATIONS}
          viewer={VIEWER}
          langs={['es', 'ca']}
          fSelected={false}
          changeSelectedKey={() => {}}
          styleKeyCol={STYLE_KEY_COL}
          styleLangCol={STYLE_LANG_COL}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('02 key with translations', () => {
    const tree = renderer
      .create(
        <HoverableTranslatorRow
          theKey={KEY_WITH_TRANSLATIONS}
          viewer={VIEWER}
          langs={['es', 'ca']}
          fSelected={false}
          changeSelectedKey={() => {}}
          styleKeyCol={STYLE_KEY_COL}
          styleLangCol={STYLE_LANG_COL}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('03 selected key', () => {
    const tree = renderer
      .create(
        <HoverableTranslatorRow
          theKey={KEY_WITHOUT_TRANSLATIONS}
          viewer={VIEWER}
          langs={['es', 'ca']}
          fSelected
          changeSelectedKey={() => {}}
          styleKeyCol={STYLE_KEY_COL}
          styleLangCol={STYLE_LANG_COL}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('04 click changes selected key', () => {
    const spyChangeSelectedKey = jest.fn();
    const tree = renderer
      .create(
        <HoverableTranslatorRow
          theKey={KEY_WITHOUT_TRANSLATIONS}
          viewer={VIEWER}
          langs={['es', 'ca']}
          fSelected={false}
          changeSelectedKey={spyChangeSelectedKey}
          styleKeyCol={STYLE_KEY_COL}
          styleLangCol={STYLE_LANG_COL}
        />
      )
      .toJSON();
    tree.props.onClick();
    expect(spyChangeSelectedKey).toBeCalledWith(KEY_WITHOUT_TRANSLATIONS.id);
  });
});
