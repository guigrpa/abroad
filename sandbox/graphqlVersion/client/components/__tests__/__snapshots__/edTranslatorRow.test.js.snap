// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoverableTranslatorRow 01 key without translations 1`] = `
<div
  className="tableBodyRow"
  id="keyId1"
  onClick={[Function]}
  style={
    Object {
      "WebkitFlex": "none",
      "display": "flex",
      "flex": "none",
      "flexDirection": "row",
      "minHeight": 21,
    }
  }
>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    style={
      Object {
        "WebkitFlex": "1 1 0px",
        "backgroundColor": "white",
        "borderBottom": "1px solid #c6bdff",
        "flex": "1 1 0px",
        "marginRight": 5,
        "paddingBottom": 1,
        "paddingLeft": 5,
        "paddingRight": 17,
        "paddingTop": 1,
        "position": "relative",
      }
    }
  >
    <span
      style={
        Object {
          "fontWeight": 900,
          "marginRight": 10,
        }
      }
    >
      someContext
    </span>
    <span
      style={
        Object {
          "whiteSpace": "pre-wrap",
        }
      }
    >
      A message
    </span>
  </div>
  <div
    style={
      Object {
        "WebkitFlex": "1 1 0px",
        "backgroundColor": "#d6e4ff",
        "borderBottom": "1px solid #c6bdff",
        "flex": "1 1 0px",
        "marginRight": 5,
        "paddingBottom": 1,
        "paddingLeft": 5,
        "paddingRight": 5,
        "paddingTop": 1,
        "position": "relative",
      }
    }
  >
    <div
      changeSelectedKey={[Function]}
      dataMockType="Translation"
      lang="es"
      theKey={
        Object {
          "context": "someContext",
          "id": "keyId1",
          "text": "A message",
          "translations": Object {
            "edges": Array [],
          },
          "unusedSince": null,
        }
      }
      translation={null}
    >
      Translation
    </div>
  </div>
  <div
    style={
      Object {
        "WebkitFlex": "1 1 0px",
        "backgroundColor": "#d6e4ff",
        "borderBottom": "1px solid #c6bdff",
        "flex": "1 1 0px",
        "marginRight": 5,
        "paddingBottom": 1,
        "paddingLeft": 5,
        "paddingRight": 5,
        "paddingTop": 1,
        "position": "relative",
      }
    }
  >
    <div
      changeSelectedKey={[Function]}
      dataMockType="Translation"
      lang="ca"
      theKey={
        Object {
          "context": "someContext",
          "id": "keyId1",
          "text": "A message",
          "translations": Object {
            "edges": Array [],
          },
          "unusedSince": null,
        }
      }
      translation={null}
    >
      Translation
    </div>
  </div>
</div>
------------HTML PREVIEW---------------
<div
  class="tableBodyRow"
  id="keyId1"
  style="-webkit-flex:none;display:flex;flex:none;flex-direction:row;min-height:21px;"
>
  <div style="-webkit-flex:1 1 0px;background-color:white;border-bottom:1px solid #c6bdff;flex:1 1 0px;margin-right:5px;padding-bottom:1px;padding-left:5px;padding-right:17px;padding-top:1px;position:relative;">
    <span style="font-weight:900;margin-right:10px;">
      someContext
    </span>
    <span style="white-space:pre-wrap;">
      A message
    </span>
  </div>
  <div style="-webkit-flex:1 1 0px;background-color:#d6e4ff;border-bottom:1px solid #c6bdff;flex:1 1 0px;margin-right:5px;padding-bottom:1px;padding-left:5px;padding-right:5px;padding-top:1px;position:relative;">
    <div
      dataMockType="Translation"
      lang="es"
    >
      Translation
    </div>
  </div>
  <div style="-webkit-flex:1 1 0px;background-color:#d6e4ff;border-bottom:1px solid #c6bdff;flex:1 1 0px;margin-right:5px;padding-bottom:1px;padding-left:5px;padding-right:5px;padding-top:1px;position:relative;">
    <div
      dataMockType="Translation"
      lang="ca"
    >
      Translation
    </div>
  </div>
</div>
`;

exports[`HoverableTranslatorRow 02 key with translations 1`] = `
<div
  className="tableBodyRow"
  id="keyId2"
  onClick={[Function]}
  style={
    Object {
      "WebkitFlex": "none",
      "display": "flex",
      "flex": "none",
      "flexDirection": "row",
      "minHeight": 21,
    }
  }
>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    style={
      Object {
        "WebkitFlex": "1 1 0px",
        "backgroundColor": "white",
        "borderBottom": "1px solid #c6bdff",
        "flex": "1 1 0px",
        "marginRight": 5,
        "paddingBottom": 1,
        "paddingLeft": 5,
        "paddingRight": 17,
        "paddingTop": 1,
        "position": "relative",
      }
    }
  >
    <span
      style={
        Object {
          "fontWeight": 900,
          "marginRight": 10,
        }
      }
    >
      someContext
    </span>
    <span
      style={
        Object {
          "whiteSpace": "pre-wrap",
        }
      }
    >
      A message
    </span>
  </div>
  <div
    style={
      Object {
        "WebkitFlex": "1 1 0px",
        "backgroundColor": "white",
        "borderBottom": "1px solid #c6bdff",
        "flex": "1 1 0px",
        "marginRight": 5,
        "paddingBottom": 1,
        "paddingLeft": 5,
        "paddingRight": 5,
        "paddingTop": 1,
        "position": "relative",
      }
    }
  >
    <div
      changeSelectedKey={[Function]}
      dataMockType="Translation"
      lang="es"
      theKey={
        Object {
          "context": "someContext",
          "id": "keyId2",
          "text": "A message",
          "translations": Object {
            "edges": Array [
              Object {
                "node": Object {
                  "id": "translationId1",
                  "lang": "es",
                  "translation": "Un mensaje",
                },
              },
              Object {
                "node": Object {
                  "id": "translationId2",
                  "lang": "ca",
                  "translation": "Un missatge",
                },
              },
            ],
          },
          "unusedSince": null,
        }
      }
      translation={
        Object {
          "id": "translationId1",
          "lang": "es",
          "translation": "Un mensaje",
        }
      }
    >
      Translation
    </div>
  </div>
  <div
    style={
      Object {
        "WebkitFlex": "1 1 0px",
        "backgroundColor": "white",
        "borderBottom": "1px solid #c6bdff",
        "flex": "1 1 0px",
        "marginRight": 5,
        "paddingBottom": 1,
        "paddingLeft": 5,
        "paddingRight": 5,
        "paddingTop": 1,
        "position": "relative",
      }
    }
  >
    <div
      changeSelectedKey={[Function]}
      dataMockType="Translation"
      lang="ca"
      theKey={
        Object {
          "context": "someContext",
          "id": "keyId2",
          "text": "A message",
          "translations": Object {
            "edges": Array [
              Object {
                "node": Object {
                  "id": "translationId1",
                  "lang": "es",
                  "translation": "Un mensaje",
                },
              },
              Object {
                "node": Object {
                  "id": "translationId2",
                  "lang": "ca",
                  "translation": "Un missatge",
                },
              },
            ],
          },
          "unusedSince": null,
        }
      }
      translation={
        Object {
          "id": "translationId2",
          "lang": "ca",
          "translation": "Un missatge",
        }
      }
    >
      Translation
    </div>
  </div>
</div>
------------HTML PREVIEW---------------
<div
  class="tableBodyRow"
  id="keyId2"
  style="-webkit-flex:none;display:flex;flex:none;flex-direction:row;min-height:21px;"
>
  <div style="-webkit-flex:1 1 0px;background-color:white;border-bottom:1px solid #c6bdff;flex:1 1 0px;margin-right:5px;padding-bottom:1px;padding-left:5px;padding-right:17px;padding-top:1px;position:relative;">
    <span style="font-weight:900;margin-right:10px;">
      someContext
    </span>
    <span style="white-space:pre-wrap;">
      A message
    </span>
  </div>
  <div style="-webkit-flex:1 1 0px;background-color:white;border-bottom:1px solid #c6bdff;flex:1 1 0px;margin-right:5px;padding-bottom:1px;padding-left:5px;padding-right:5px;padding-top:1px;position:relative;">
    <div
      dataMockType="Translation"
      lang="es"
    >
      Translation
    </div>
  </div>
  <div style="-webkit-flex:1 1 0px;background-color:white;border-bottom:1px solid #c6bdff;flex:1 1 0px;margin-right:5px;padding-bottom:1px;padding-left:5px;padding-right:5px;padding-top:1px;position:relative;">
    <div
      dataMockType="Translation"
      lang="ca"
    >
      Translation
    </div>
  </div>
</div>
`;

exports[`HoverableTranslatorRow 03 selected key 1`] = `
<div
  className="tableBodyRow"
  id="keyId1"
  onClick={[Function]}
  style={
    Object {
      "WebkitFlex": "none",
      "display": "flex",
      "flex": "none",
      "flexDirection": "row",
      "minHeight": 21,
    }
  }
>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    style={
      Object {
        "WebkitFlex": "1 1 0px",
        "backgroundColor": "#dcd6ff",
        "borderBottom": "1px solid #c6bdff",
        "flex": "1 1 0px",
        "marginRight": 5,
        "paddingBottom": 1,
        "paddingLeft": 5,
        "paddingRight": 17,
        "paddingTop": 1,
        "position": "relative",
      }
    }
  >
    <span
      style={
        Object {
          "fontWeight": 900,
          "marginRight": 10,
        }
      }
    >
      someContext
    </span>
    <span
      style={
        Object {
          "whiteSpace": "pre-wrap",
        }
      }
    >
      A message
    </span>
  </div>
  <div
    style={
      Object {
        "WebkitFlex": "1 1 0px",
        "backgroundColor": "#dcd6ff",
        "borderBottom": "1px solid #c6bdff",
        "flex": "1 1 0px",
        "marginRight": 5,
        "paddingBottom": 1,
        "paddingLeft": 5,
        "paddingRight": 5,
        "paddingTop": 1,
        "position": "relative",
      }
    }
  >
    <div
      changeSelectedKey={[Function]}
      dataMockType="Translation"
      lang="es"
      theKey={
        Object {
          "context": "someContext",
          "id": "keyId1",
          "text": "A message",
          "translations": Object {
            "edges": Array [],
          },
          "unusedSince": null,
        }
      }
      translation={null}
    >
      Translation
    </div>
  </div>
  <div
    style={
      Object {
        "WebkitFlex": "1 1 0px",
        "backgroundColor": "#dcd6ff",
        "borderBottom": "1px solid #c6bdff",
        "flex": "1 1 0px",
        "marginRight": 5,
        "paddingBottom": 1,
        "paddingLeft": 5,
        "paddingRight": 5,
        "paddingTop": 1,
        "position": "relative",
      }
    }
  >
    <div
      changeSelectedKey={[Function]}
      dataMockType="Translation"
      lang="ca"
      theKey={
        Object {
          "context": "someContext",
          "id": "keyId1",
          "text": "A message",
          "translations": Object {
            "edges": Array [],
          },
          "unusedSince": null,
        }
      }
      translation={null}
    >
      Translation
    </div>
  </div>
</div>
------------HTML PREVIEW---------------
<div
  class="tableBodyRow"
  id="keyId1"
  style="-webkit-flex:none;display:flex;flex:none;flex-direction:row;min-height:21px;"
>
  <div style="-webkit-flex:1 1 0px;background-color:#dcd6ff;border-bottom:1px solid #c6bdff;flex:1 1 0px;margin-right:5px;padding-bottom:1px;padding-left:5px;padding-right:17px;padding-top:1px;position:relative;">
    <span style="font-weight:900;margin-right:10px;">
      someContext
    </span>
    <span style="white-space:pre-wrap;">
      A message
    </span>
  </div>
  <div style="-webkit-flex:1 1 0px;background-color:#dcd6ff;border-bottom:1px solid #c6bdff;flex:1 1 0px;margin-right:5px;padding-bottom:1px;padding-left:5px;padding-right:5px;padding-top:1px;position:relative;">
    <div
      dataMockType="Translation"
      lang="es"
    >
      Translation
    </div>
  </div>
  <div style="-webkit-flex:1 1 0px;background-color:#dcd6ff;border-bottom:1px solid #c6bdff;flex:1 1 0px;margin-right:5px;padding-bottom:1px;padding-left:5px;padding-right:5px;padding-top:1px;position:relative;">
    <div
      dataMockType="Translation"
      lang="ca"
    >
      Translation
    </div>
  </div>
</div>
`;
